### ISSUE TITLE
Listing IPFS hash versioning

### ISSUE THREAD

**Issue by tyleryasaka** [OriginProtocol/origin-js#250]
In our [fractional usage discussion](https://github.com/OriginProtocol/origin-js/issues/146), we've decided that the IPFS hashes should be updatable and versioned. Since we're going to have a base listing contract, I think the versioning should happen there (whether our dapp supports IPFS data updates for other listing types or not).

I think the simplest way to implement this right now is to have an array of IPFS hashes on the base listing contract. The last item of the array is considered the most recent version. ~We could eventually get fancier and put timestamps on the versions, but I don't think that's necessary yet.~ EDIT: I think we should go ahead and be fancy and add timestamps. The reason we're adding versioning in the first place is primarily arbitration. Timestamps seem pretty essential to arbitration and also go naturally with versions, so we should probably go ahead and include them.

**Comment by nick** [comment]
I think we should try to avoid storing anything that can be determined by other methods so we're not using contract storage unless absolutely necessary. Using a single value instead of an array for ipfsHash will also reduce contract complexity. 

Instead, we could emit a ListingUpdated(bytes32 ipfsHash) event whenever a listing is updated. We could then use the [getPastEvents](http://web3js.readthedocs.io/en/1.0/web3-eth-contract.html#getpastevents) method from web3 filtering on ListingUpdated to give us the ipfsHash history. The returned events also have a blocknumber property which we can use with [getBlock](http://web3js.readthedocs.io/en/1.0/web3-eth.html#getblock) to determine the timestamp.

**Comment by tyleryasaka** [comment]
@nick Fair point.

That said, I realized that if we make listings editable, we'll need to deal with this issue: https://github.com/OriginProtocol/origin-js/issues/19

TLDR: purchase contracts should reference specific versions of listing contracts, if listing contracts are editable.

We can still rely on events to get previous IPFS hashes. However, a couple of concerns:
1. If at any point in the future we want anything editable to be stored directly on-chain (as opposed to in IPFS), then we'll need to store versions explicitly on-chain.
1. I wonder if it could be ambiguous or confusing which IPFS hash a purchase contract points to, if we're just looking at the block number of the event? If we stored versions on-chain, we could have the purchase contract programmatically save the version number of the listing that it points to. If we're just looking at events, we'll have to infer what the active version of the listing was at the time of purchase contract creation.

With those thoughts in mind, I'm not sure it's worth trying to optimize this part right now. I would think the first step towards optimization would be moving purchase contract code to a library.
